notebooks/generative-ai/question-answering.ipynb

```diff
@@ -208,43 +244,375 @@
+        "from langchain.schema.output_parser import StrOutputParser\n",
+        "\n",
+        "retriever = vector_store.as_retriever()\n",
+        "\n",
+        "llm = OpenAI(openai_api_key=OPENAI_API_KEY)\n",
+        "\n",
+        "ANSWER_PROMPT = ChatPromptTemplate.from_template(\n",
+        "    \"\"\"Answer the question based only on the following context:\n",
+        "    {context}\n",
+        "\n",
+        "    Question: {question}\n",
+        "    \"\"\"\n",
+        ")\n",
+        "\n",
+        "chain = (\n",
+        "    {\"context\": retriever, \"question\": RunnablePassthrough()}\n",
+        "    | ANSWER_PROMPT\n",
+        "    | llm\n",
+        "    | StrOutputParser()\n",
+        ")\n",
+        "\n",
+        "ans = chain.invoke(\"what is the nasa sales team?\")\n",
+        "\n",
+        "print(\"---- Answer ----\")\n",
+        "print(ans)"
+      ]
+    },
+    {
+      "cell_type": "markdown",
+      "metadata": {},
+      "source": [
+        "### Add Source Tracing\n",
+        "RAG can provide clear traceability of the source knowledge used to answer a question. This is important for compliance and regulatory reasons and limiting LLM hallucinations. This is known as source tracking.\n",
+        "\n",
+        "In this example, we extend the Prompt template to ask the LLM to cite the source of the answer."
+      ]
+    },
+    {
+      "cell_type": "code",
+      "execution_count": 26,
+      "metadata": {},
+      "outputs": [
+        {
+          "name": "stdout",
+          "output_type": "stream",
+          "text": [
+            "---- Answer ----\n",
+            "\n",
+            "Answer: The NASA Sales Team is the North America South America Region (NASA) sales team. This team is led by two Area Vice-Presidents: Laura Martinez is the Area Vice-President of North America, and Gary Johnson is the Area Vice-President of South America. \n",
+            "\n",
+            "Source: Sales Organization Overview\n"
+          ]
+        }
+      ],
+      "source": [
+        "from langchain.schema.runnable import RunnablePassthrough\n",
+        "from langchain.prompts import ChatPromptTemplate, PromptTemplate\n",
+        "from langchain.schema.output_parser import StrOutputParser\n",
+        "from langchain.schema import format_document\n",
+        "\n",
+        "retriever = vector_store.as_retriever()\n",
+        "\n",
+        "llm = OpenAI(openai_api_key=OPENAI_API_KEY)\n",
+        "\n",
+        "ANSWER_PROMPT = ChatPromptTemplate.from_template(\n",
+        "\"\"\"\n",
+        "Use the following passages to answer the user's question.\n",
+        "Each passage has a SOURCE which is the title of the document. When answering, cite source name of the passages you are answering from below the answer in a unique bullet point list.\n",
+        "\n",
+        "If you don't know the answer, just say that you don't know, don't try to make up an answer.\n",
```

Comment:
In your experience, does this help reduce hallucinations? Isn't it a bit like asking a drunk person to act sober?

Reply:
Yes, you need to instruct this otherwise the LLM will try and use its own knowledge to answer the question, if its not in the context and likely make up an answer.

Reply:
Then wouldn't you say somthing like "if the question isn't answered in the context information, say I don't know; don't try to make up an answer from other sources".

Reply:
Yeh - being terse as smaller the prompt, less cost to generate. Happy to discuss later on optimizing the prompt. IMO it will be something i would expect the developer to tweak.

Reply:
My issue with this prompt is that it implies that the AI knows what it knows and what it doesn't, and is able to make decisions based on that, which is false. All the AI can do is find sequences of words that are probabilistically more likely to agree with the prompt based on what it has seen during training. "If you don't know the answer..." is very vague in that respect, something like "If the answer isn't in the context provided..." is in my opinion a much better guidance.